### PR TITLE
[openmp-extras] Working OpenMP offloading v 4.3.1

### PR DESCRIPTION
**Pull request by linedot**
OpenMP offloading works in C. No flang.

### CODE REVIEW

**Comment by linedot** [comment]
Updated for 4.5.0 and made into a combined package for openmp-extras, rocm-pgmath, rocm-flang and rocm-flang-runtime

File: openmp-extras/PKGBUILD [hunk 1]
```diff
@@ -1,12 +1,121 @@
...
 arch=('x86_64')
-url="https://rocmdocs.amd.com/en/latest/Programming_Guides/aomp.html#openmp-support"
-license=()
-depends=('llvm-amdgpu' 'rocm-device-libs')
-makedepends=()
-source=()
-sha256sums=()
+url='https://github.com/ROCm-Developer-Tools/aomp'
+license=('custom:Apache 2.0 with LLVM Exception')
+depends=(rocm-device-libs llvm-amdgpu)
```
Comment (tpkessler) [inline openmp-extras/PKGBUILD:9]: llvm-amdgpu is now called rocm-llvm

File: openmp-extras/PKGBUILD [hunk 2]
```diff
@@ -1,12 +1,121 @@
...
+}
+
+
+package_openmp-extras() {
+
+    mkdir -p $pkgdir/opt/rocm/
+    cp -r $srcdir/out/openmp-extras/* $pkgdir/opt/rocm/
+}
+
+package_rocm-pgmath() {
```
Comment (tpkessler) [inline openmp-extras/PKGBUILD:105]: We try to follow upstream packages as close as possible. In the Debian package, the pgmath/flang part is part of openmp-extras.

File: openmp-extras/aomp-flang-decouple-out-dir-and-rocm-install.patch [hunk 3]
```diff
@@ -0,0 +1,57 @@
+diff --color -urN aomp-rocm-4.5.0-unpatched/bin/build_flang_runtime.sh aomp-rocm-4.5.0/bin/build_flang_runtime.sh
```
Comment (tpkessler) [inline openmp-extras/aomp-flang-decouple-out-dir-and-rocm-install.patch:1]: Great work! I always had issues with this part and it stopped me from working on the package. Thank you.

**Review by tpkessler — CHANGES_REQUESTED** [review]
Great work! I greatly appreciate it. Before we can merge please address my comments on the PKGBUILD.

File: openmp-extras/PKGBUILD (hunk 2 above)
Comment (linedot) [reply openmp-extras/PKGBUILD:105]: The fortran compiler being part of an "openmp" package just felt weird to me, but I'll change this. Makes the PKGBUILD simpler even.

File: openmp-extras/aomp-flang-decouple-out-dir-and-rocm-install.patch (hunk 3 above)
Comment (linedot) [reply openmp-extras/aomp-flang-decouple-out-dir-and-rocm-install.patch:1]: Yeah the whole "build and install step-by-step directly into the filesystem" approach AMD is taking is a bit awkward

File: openmp-extras/PKGBUILD (hunk 1 above)
Comment (linedot) [reply openmp-extras/PKGBUILD]: changed

**Comment by tpkessler** [comment]
Could you please update SRCINFO?

**Comment by linedot** [comment]
done

**Comment by tpkessler** [comment]
Did you manage to compile `openmp-extras`? I had to add additional dependencies and a new patch for flang,

```patch
diff --git a/openmp-extras/PKGBUILD b/openmp-extras/PKGBUILD
index 1594f7f..f025a75 100644
--- a/openmp-extras/PKGBUILD
+++ b/openmp-extras/PKGBUILD
@@ -5,8 +5,8 @@ pkgrel=1
 arch=('x86_64')
 url='https://github.com/ROCm-Developer-Tools/aomp'
 license=('custom:Apache 2.0 with LLVM Exception')
-depends=(rocm-device-libs rocm-llvm)
-makedepends=(cmake python ninja gawk)
+depends=(hsa-rocr rocm-llvm)
+makedepends=(cmake python python-pip python-wheel ninja gawk)
 llvm_url='https://github.com/RadeonOpenCompute/llvm-project'
 dlibs_url='https://github.com/RadeonOpenCompute/ROCm-Device-Libs'
 flang_url='https://github.com/ROCm-Developer-Tools/flang'
@@ -18,6 +18,7 @@ source=("llvm-amdgpu-${pkgver}.tar.gz::$llvm_url/archive/rocm-$pkgver.tar.gz"
         "aomp-openmp-buildpath.patch"
         "aomp-flang-decouple-out-dir-and-rocm-install.patch"
         "aomp-flang-libomp-path.patch"
+        "aomp-flang-remove-sha.patch"
         )
 sha256sums=('e33d08c8ae42a3ba394dceb8938263cd14ba45e9603b18f3502c3344814ca296'
             '27a5794b5885c61dc6f63cec36673b37deb029754d3b2fd3e1b21239efffa96a'
@@ -27,6 +28,7 @@ sha256sums=('e33d08c8ae42a3ba394dceb8938263cd14ba45e9603b18f3502c3344814ca296'
             'f7ed1704ffb095bbe8512b1c567a111936685d35f64123c786194e4239277251'
             'e82a4f065cc259095bf96778b913b97fe39d7c207e4e25ccf59d8fa264014262'
             'ff3c3e56bfc11c0c9a6ab5c5392168da06aed3b4a21cdfcf7a149d12a69e2906'
+            'dcb991033f79643bdb20b04d1174ba4519ede664aebb1eba15a058c887e4665e'
             )
 options=(staticlibs)

@@ -40,6 +42,8 @@ prepare() {
     patch -Np1 < $srcdir/aomp-openmp-buildpath.patch
     patch -Np1 < $srcdir/aomp-flang-decouple-out-dir-and-rocm-install.patch
     patch -Np1 < $srcdir/aomp-flang-libomp-path.patch
+    cd $srcdir/flang
+    patch -Np1 < $srcdir/aomp-flang-remove-sha.patch

 }

```

The new patch file is
```patch
*** flang-rocm-4.5.0/CMakeLists.txt.bak	2021-11-19 18:08:01.938702647 +0100
--- flang-rocm-4.5.0/CMakeLists.txt	2021-11-19 18:08:11.666666963 +0100
***************
*** 375,384 ****
    add_definitions( -DFLANG_VERSION_MAJOR="${FLANG_VERSION_MAJOR} " )
    add_definitions( -DFLANG_VERSION_MINOR="${FLANG_VERSION_MINOR} " )
  endif()
- execute_process(COMMAND git log -1 --format=format:%H WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} OUTPUT_VARIABLE FLANG_SHA)
- string(STRIP ${FLANG_SHA} FLANG_SHA)
- message(STATUS "MY_VAR=${FLANG_SHA}")
- add_definitions( -DFLANG_SHA="Flang Sha ${FLANG_SHA} ")

  option(FLANG_INCLUDE_TESTS
         "Generate build targets for the Flang unit tests."
--- 375,380 ----
```

**Comment by tpkessler** [comment]
I found the issue. The problem was that I removed `FLANG_SHA` but it is referred to later in the code. The updated patch is
```patch
*** flang-rocm-4.5.0/CMakeLists.txt.bak	2021-11-19 18:39:54.679976196 +0100
--- flang-rocm-4.5.0/CMakeLists.txt	2021-11-19 18:40:12.911889722 +0100
***************
*** 375,384 ****
    add_definitions( -DFLANG_VERSION_MAJOR="${FLANG_VERSION_MAJOR} " )
    add_definitions( -DFLANG_VERSION_MINOR="${FLANG_VERSION_MINOR} " )
  endif()
! execute_process(COMMAND git log -1 --format=format:%H WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR} OUTPUT_VARIABLE FLANG_SHA)
! string(STRIP ${FLANG_SHA} FLANG_SHA)
! message(STATUS "MY_VAR=${FLANG_SHA}")
! add_definitions( -DFLANG_SHA="Flang Sha ${FLANG_SHA} ")

  option(FLANG_INCLUDE_TESTS
         "Generate build targets for the Flang unit tests."
--- 375,381 ----
    add_definitions( -DFLANG_VERSION_MAJOR="${FLANG_VERSION_MAJOR} " )
    add_definitions( -DFLANG_VERSION_MINOR="${FLANG_VERSION_MINOR} " )
  endif()
! add_definitions( -DFLANG_SHA="Flang Sha None ")

  option(FLANG_INCLUDE_TESTS
         "Generate build targets for the Flang unit tests."
``` 

With  this patch, the package builds in a clean chroot.

**Comment by tpkessler** [comment]
If I however apply it to simple example it fails. With test.c
```C
#include <stdio.h>
#include <omp.h>

int main(int argc, char *argv[])
{
    #pragma omp target
    printf("I'm running on the %s.\n", omp_is_initial_device() ? "host" : "accelerator");

    return 0;
}
```
I get
```
hipcc -target amdgcn-amd-amdhsa -fopenmp -fopenmp-targets=amdgcn-amd-amdhsa -Xopenmp-target=amdgcn-amd-amdhsa -march=gfx900 test.c

clang-13: error: no such file or directory: 'libomptarget-amdgcn-gfx900.bc'
clang-13: error: no such file or directory: 'libaompextras-amdgcn-gfx900.bc'
```
If I list all `*.bc`, they are not there
```
pacman -Ql openmp-extras| grep bc

openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libbc-hostrpc-amdgcn.a
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx1030.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx1031.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx700.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx701.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx801.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx803.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx900.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx902.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx906.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx908.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx90a.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-nvptx-sm_30.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-nvptx-sm_35.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-nvptx-sm_50.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-nvptx-sm_60.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-nvptx-sm_61.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-nvptx-sm_70.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx1030.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx1031.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx700.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx701.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx801.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx803.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx900.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx902.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx906.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx908.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx90a.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libbc-hostrpc-amdgcn.a
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx1030.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx1031.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx700.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx701.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx801.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx803.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx900.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx902.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx906.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx908.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx90a.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-nvptx-sm_30.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-nvptx-sm_35.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-nvptx-sm_50.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-nvptx-sm_60.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-nvptx-sm_61.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-nvptx-sm_70.bc
```

The upstream package, however, contains the files we are missing.

**Comment by linedot** [comment]
Hi, the package builds cleanly for me without the patch, I'm not building in a clean chroot though, not sure how it affects things.

OpenMP works for me with /opt/rocm/llvm/bin/clang*. I am using a gfx1030 though. here's my example:
```
#include <vector>
#include <random>
#include <algorithm>
#include <fmt/core.h>


constexpr std::uint64_t vector_size = 1024;

int main()
{
    std::vector<double> avec(vector_size);
    std::vector<double> bvec(vector_size);
    std::vector<double> cvec(vector_size);


    std::random_device rnd;
    std::vector<std::uint64_t> seed_data(624);
    std::generate(begin(seed_data), end(seed_data),[&](){return rnd();});
    std::seed_seq seed(begin(seed_data), end(seed_data));

    std::mt19937_64 rng(seed);

    std::uniform_real_distribution<double> dist(0.0,1.0);

    std::generate(begin(avec), end(avec), [&](){return dist(rng);});
    std::generate(begin(bvec), end(bvec), [&](){return dist(rng);});
    std::generate(begin(cvec), end(cvec), [&](){return dist(rng);});


    double* a = avec.data();
    double* b = bvec.data();
    double* c = cvec.data();


    fmt::print("Calculating vector FMA on GPU\n");

    #pragma omp target teams map(to: a[0:vector_size],b[0:vector_size]) map(from:c[0:vector_size])
    #pragma omp distribute parallel for
    for(std::uint64_t i = 0; i < vector_size; i++)
    {
        c[i] = a[i]*b[i]+c[i];
    }

    fmt::print("Calculation successfull\n");

    return 0;
}
```
Building:
```
/opt/rocm/llvm/bin/clang++ -O3 -target x86_64-pc-linux-gnu -fopenmp -fopenmp-targets=amdgcn-amd-amdhsa -Xopenmp-target=amdgcn-amd-amdhsa -march=gfx1030 -fPIC -o vectoradd.cpp.o -c vectoradd.cpp
/opt/rocm/llvm/bin/clang++ -target x86_64-pc-linux-gnu -fopenmp -fopenmp-targets=amdgcn-amd-amdhsa -Xopenmp-target=amdgcn-amd-amdhsa -march=gfx1030  vectoradd.cpp.o -o vectoradd -lfmt
```
Running:
```
 ᐅ LIBOMPTARGET_KERNEL_TRACE=1  ./vectoradd     
Calculating vector FMA on GPU
DEVID: 0 SGN:0 ConstWGSize:256  args: 3 teamsXthrds:(   4X 256) reqd:(   0X   0) sgpr_count:34 vgpr_count:46 sgpr_spill_count:0 vgpr_spill_count:0 tripcount:1024 n:__omp_offloading_2d_1b65971_main_l37
Calculation successfull
```

Your example:
```
~/SourceCode/test/ROCM/OpenMP ᐅ /opt/rocm/llvm/bin/clang -target x86_64-pc-linux-gnu -fopenmp -fopenmp-targets=amdgcn-amd-amdhsa -Xopenmp-target=amdgcn-amd-amdhsa -march=gfx1030 omptargettest.c -o omptargettest
~/SourceCode/test/ROCM/OpenMP ᐅ LIBOMPTARGET_KERNEL_TRACE=1  ./omptargettest 
DEVID: 0 SGN:1 ConstWGSize:256  args: 0 teamsXthrds:(   1X 256) reqd:(   1X   0) sgpr_count:36 vgpr_count:80 sgpr_spill_count:0 vgpr_spill_count:0 tripcount:0 n:__omp_offloading_2d_27c03fb_main_l6
I'm running on the accelerator.
```

**Comment by linedot** [comment]
```
ᐅ pacman -Ql openmp-extras| grep bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libbc-hostrpc-amdgcn.a
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx1030.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx1031.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx700.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx701.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx801.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx803.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx900.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx902.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx906.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx908.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-amdgcn-gfx90a.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-nvptx-sm_30.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-nvptx-sm_35.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-nvptx-sm_50.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-nvptx-sm_60.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-nvptx-sm_61.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libm-nvptx-sm_70.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx1030.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx1031.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx700.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx701.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx801.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx803.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx900.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx902.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx906.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx908.bc
openmp-extras /opt/rocm/llvm/lib-debug/libdevice/libomptarget-amdgcn-gfx90a.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libbc-hostrpc-amdgcn.a
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx1030.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx1031.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx700.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx701.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx801.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx803.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx900.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx902.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx906.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx908.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-amdgcn-gfx90a.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-nvptx-sm_30.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-nvptx-sm_35.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-nvptx-sm_50.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-nvptx-sm_60.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-nvptx-sm_61.bc
openmp-extras /opt/rocm/llvm/lib/libdevice/libm-nvptx-sm_70.bc
```

**Comment by linedot** [comment]
whoops:
```
find /opt/rocm -name libaompextras-amdgcn\*
/opt/rocm/llvm/lib/libdevice/libaompextras-amdgcn-gfx1030.bc
/opt/rocm/llvm/lib/libdevice/libaompextras-amdgcn-gfx1031.bc
/opt/rocm/llvm/lib/libdevice/libaompextras-amdgcn-gfx700.bc
/opt/rocm/llvm/lib/libdevice/libaompextras-amdgcn-gfx701.bc
/opt/rocm/llvm/lib/libdevice/libaompextras-amdgcn-gfx801.bc
/opt/rocm/llvm/lib/libdevice/libaompextras-amdgcn-gfx803.bc
/opt/rocm/llvm/lib/libdevice/libaompextras-amdgcn-gfx900.bc
/opt/rocm/llvm/lib/libdevice/libaompextras-amdgcn-gfx902.bc
/opt/rocm/llvm/lib/libdevice/libaompextras-amdgcn-gfx906.bc
/opt/rocm/llvm/lib/libdevice/libaompextras-amdgcn-gfx908.bc
/opt/rocm/llvm/lib/libdevice/libaompextras-amdgcn-gfx90a.bc
```
```
pacman -Qo /opt/rocm/llvm/lib/libdevice/libaompextras-amdgcn-gfx1030.bc
error: No package owns /opt/rocm/llvm/lib/libdevice/libaompextras-amdgcn-gfx1030.bc
```
I need to figure out how I got these files.

**Comment by linedot** [comment]
How about this? (Also added your flang sha patch)

**Comment by tpkessler** [comment]
That's looking good! But we have to add yet another patch, this time for the LLVM path used by openmp.

```patch
diff --git a/openmp-extras/PKGBUILD b/openmp-extras/PKGBUILD
index bb2ffd7..94f0fd9 100644
--- a/openmp-extras/PKGBUILD
+++ b/openmp-extras/PKGBUILD
@@ -18,6 +18,7 @@ source=("llvm-amdgpu-${pkgver}.tar.gz::$llvm_url/archive/rocm-$pkgver.tar.gz"
         "flang-$pkgver.tar.gz::$flang_url/archive/rocm-$pkgver.tar.gz"
         "aomp-gcc8-only-for-cuda.patch"
         "aomp-openmp-buildpath.patch"
+        "aomp-openmp-llvm.patch"
         "aomp-flang-decouple-out-dir-and-rocm-install.patch"
         "aomp-flang-libomp-path.patch"
         "aomp-flang-remove-sha.patch"
@@ -30,6 +31,7 @@ sha256sums=('e33d08c8ae42a3ba394dceb8938263cd14ba45e9603b18f3502c3344814ca296'
             '3d7277fd658e51e7e43272c4b319e733c18f5a6d11f739aaec0df758a720809e'
             '705a7103c3aeff514e5645c130786172961c54673dfdd772caece3b5e7536088'
             'f7ed1704ffb095bbe8512b1c567a111936685d35f64123c786194e4239277251'
+            'dfb76e7b796f3e75e752efa2a4cb8157accd7f324e985e63a0dc43f275edae9f'
             'e82a4f065cc259095bf96778b913b97fe39d7c207e4e25ccf59d8fa264014262'
             'ff3c3e56bfc11c0c9a6ab5c5392168da06aed3b4a21cdfcf7a149d12a69e2906'
             '565af54e3079c130c87ebe036d91db21631e01ff0435f80bc54dbb55e8b02229'
@@ -46,6 +48,7 @@ prepare() {
     cd $srcdir/aomp-rocm-$pkgver
     patch -Np1 < $srcdir/aomp-gcc8-only-for-cuda.patch
     patch -Np1 < $srcdir/aomp-openmp-buildpath.patch
+    patch -Np1 < $srcdir/aomp-openmp-llvm.patch
     patch -Np1 < $srcdir/aomp-flang-decouple-out-dir-and-rocm-install.patch
     patch -Np1 < $srcdir/aomp-flang-libomp-path.patch

```

```patch
*** aomp-rocm-4.5.0/bin/build_openmp.sh.bak	2021-11-20 12:12:42.802229426 +0100
--- aomp-rocm-4.5.0/bin/build_openmp.sh	2021-11-20 12:14:46.151777013 +0100
***************
*** 125,131 ****
  -DLLVM_INSTALL_PREFIX=$OUT_DIR/llvm
  -DLLVM_MAIN_INCLUDE_DIR=$LLVM_PROJECT_ROOT/llvm/include
  -DLIBOMPTARGET_LLVM_INCLUDE_DIRS=$LLVM_PROJECT_ROOT/llvm/include
! -DCMAKE_PREFIX_PATH=$ROCM_DIR;$ROCM_DIR/include/hsa;$DEVICELIBS_BUILD_PATH;$OUT_DIR/build/devicelibs"

  if [ "$AOMP_BUILD_CUDA" == 1 ] ; then
     COMMON_CMAKE_OPTS="$COMMON_CMAKE_OPTS
--- 125,132 ----
  -DLLVM_INSTALL_PREFIX=$OUT_DIR/llvm
  -DLLVM_MAIN_INCLUDE_DIR=$LLVM_PROJECT_ROOT/llvm/include
  -DLIBOMPTARGET_LLVM_INCLUDE_DIRS=$LLVM_PROJECT_ROOT/llvm/include
! -DCMAKE_PREFIX_PATH=$ROCM_DIR;$ROCM_DIR/include/hsa;$DEVICELIBS_BUILD_PATH;$OUT_DIR/build/devicelibs
! -DLLVM_INSTALL_PREFIX=/opt/rocm/llvm"

  if [ "$AOMP_BUILD_CUDA" == 1 ] ; then
     COMMON_CMAKE_OPTS="$COMMON_CMAKE_OPTS
```

With this patch the missing `*bc` files in `lib/libdevice` are generated. In the first cmake run, LLVM is not found because it looks for it under the AOMP path. Only on the second run (for debug libraries), LLVM magically appears on the cmake output. Adding `LLVM_INSTALL_PREFIX` solves this issue. I'm now able to compile and run your vectoradd example on my Vega 56.

**Comment by linedot** [comment]
Added the patch

**Comment by tpkessler** [comment]
Thank you very much for your hard work @linedot! I'm very excited to see a fully working version `openmp-extras` in this repo.

**Pull request merged**